File: core/src/main/java/cloudspec/annotation/ResourceDefReflectionUtil.java
/*-
 * #%L
 * CloudSpec Core Library
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.annotation;

import cloudspec.model.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.Field;
import java.lang.reflect.ParameterizedType;
import java.time.Instant;
import java.util.Date;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Util to extract resource definition information out of annotated classes.
 */
public class ResourceDefReflectionUtil {
    private static final Logger LOGGER = LoggerFactory.getLogger(ResourceDefReflectionUtil.class);

    /**
     * Extract resource definition out of a class annotated with @ResourceDefinition.
     *
     * @param clazz Class.
     * @return Optional resource definition.
     */
    public static Optional<ResourceDef> toResourceDef(Class<?> clazz) {
        return toResourceDefRef(clazz).flatMap(resourceDefRef ->
                toResourceDescription(clazz).map(resourceDescription ->
                        new ResourceDef(resourceDefRef, resourceDescription, toPropertyDefs(clazz), toAssociationDefs(clazz))
                )
        );
    }

    /**
     * Extract resource definition reference out of a class annotated with @ResourceDefinition.
     *
     * @param clazz Class.
     * @return Optional resource definition reference.
     */
    public static Optional<ResourceDefRef> toResourceDefRef(Class<?> clazz) {
        if (!clazz.isAnnotationPresent(ResourceDefinition.class)) {
            LOGGER.warn(
                    "Cannot produce a resource definition reference from class {} " +
                            "because it is not annotated with @ResourceDefinition",
                    clazz.getCanonicalName()
            );
            return Optional.empty();
        }

        ResourceDefinition resourceDefAnnotation = clazz.getAnnotation(ResourceDefinition.class);

        if (LOGGER.isDebugEnabled()) {
            LOGGER.debug("Obtained provider='{}', group='{}', name='{}' from class {}",
                    resourceDefAnnotation.provider(),
                    resourceDefAnnotation.group(),
                    resourceDefAnnotation.name(),
                    clazz.getCanonicalName()
            );
        }

        return Optional.of(new ResourceDefRef(
                resourceDefAnnotation.provider(),
                resourceDefAnnotation.group(),
                resourceDefAnnotation.name()
        ));
    }

    /**
     * Extract resource description out of a class annotated with @ResourceDefinition.
     *
     * @param clazz Class.
     * @return Optional resource description.
     */
    public static Optional<String> toResourceDescription(Class<?> clazz) {
        if (!clazz.isAnnotationPresent(ResourceDefinition.class)) {
            LOGGER.warn(
                    "Cannot produce resource description from class {} " +
                            "because it is not annotated with @ResourceDefinition",
                    clazz.getCanonicalName()
            );
            return Optional.empty();
        }

        ResourceDefinition resourceDefAnnotation = clazz.getAnnotation(ResourceDefinition.class);

        if (LOGGER.isDebugEnabled()) {
            LOGGER.debug("Obtained description='{}' from class {}",
                    resourceDefAnnotation.description(),
                    clazz.getCanonicalName()
            );
        }

        return Optional.of(resourceDefAnnotation.description());
    }

    /**
     * Extract all property definitions out of fields annotated with @PropertyDefinition.
     *
     * @param clazz Class.
     * @return List of property definitions.
     */
    public static List<PropertyDef> toPropertyDefs(Class<?> clazz) {
        return Stream.of(clazz.getDeclaredFields())
                .map(field -> toPropertyDef(clazz, field))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toList());
    }

    private static Optional<PropertyDef> toPropertyDef(Class<?> resourceClass, Field field) {
        if (!field.isAnnotationPresent(PropertyDefinition.class)) {
            return Optional.empty();
        }

        PropertyDefinition propertyDefAnnotation = field.getAnnotation(PropertyDefinition.class);

        PropertyType propertyType = guessPropertyType(field);
        Boolean multiValued = guessMultiValued(field);

        switch (propertyType) {
            case KEY_VALUE:
            case NUMBER:
            case STRING:
            case BOOLEAN:
            case DATE:
                return Optional.of(
                        new PropertyDef(
                                propertyDefAnnotation.name(),
                                propertyDefAnnotation.description(),
                                propertyType,
                                multiValued,
                                propertyDefAnnotation.exampleValues()
                        )
                );
            case NESTED:
            default:
                Class<?> type = multiValued ?
                        (Class<?>) ((ParameterizedType) field.getGenericType()).getActualTypeArguments()[0] :
                        field.getType();

                // get property definitions
                List<PropertyDef> propertyDefs = Stream.of(type.getDeclaredFields())
                        .map(subField -> toPropertyDef(type, subField))
                        .filter(Optional::isPresent)
                        .map(Optional::get)
                        .collect(Collectors.toList());

                List<AssociationDef> associationDefs = Stream.of(type.getDeclaredFields())
                        .map(subField -> toAssociationDef(type, subField))
                        .filter(Optional::isPresent)
                        .map(Optional::get)
                        .collect(Collectors.toList());

                if (propertyDefs.size() > 0 || associationDefs.size() > 0) {
                    return Optional.of(
                            new PropertyDef(
                                    propertyDefAnnotation.name(),
                                    propertyDefAnnotation.description(),
                                    PropertyType.NESTED,
                                    multiValued,
                                    propertyDefAnnotation.exampleValues(),
                                    propertyDefs,
                                    associationDefs
                            )
                    );
                }
        }

        Class<?> realClazz = getContainedClass(field);
        LOGGER.warn("Type {} of property '{}' in class {} is not supported",
                realClazz.getCanonicalName(),
                field.getName(),
                resourceClass.getCanonicalName()
        );

        return Optional.empty();
    }

    /**
     * Extract all association definitions out of fields annotated with @AssociationDefinition.
     *
     * @param clazz Class.
     * @return List of association definitions.
     */
    public static List<AssociationDef> toAssociationDefs(Class<?> clazz) {
        return Stream.of(clazz.getDeclaredFields())
                .map(field -> toAssociationDef(clazz, field))
                .filter(Optional::isPresent)
                .map(Optional::get)
                .collect(Collectors.toList());
    }

    private static Optional<AssociationDef> toAssociationDef(Class<?> resourceClass, Field field) {
        if (!field.isAnnotationPresent(AssociationDefinition.class)) {
            return Optional.empty();
        }

        AssociationDefinition associationDefAnnotation = field.getAnnotation(AssociationDefinition.class);

        Boolean multiValued = guessMultiValued(field);
        Class<?> realClazz = getContainedClass(field);

        if (!realClazz.isAssignableFrom(String.class)) {
            LOGGER.warn("Type {} of association '{}' in class {} is not supported",
                    realClazz.getCanonicalName(),
                    field.getName(),
                    resourceClass.getCanonicalName()
            );

            return Optional.empty();
        }

        return toResourceDefRef(associationDefAnnotation.targetClass())
                .map(resourceDefRef -> new AssociationDef(
                        associationDefAnnotation.name(),
                        associationDefAnnotation.description(),
                        resourceDefRef,
                        multiValued
                ));
    }

    /**
     * Guess whether a field of a class is multivalued.
     *
     * @param field Field.
     * @return True if field is multi-valued. False otherwise.
     */
    public static Boolean guessMultiValued(Field field) {
        return field.getType().isAssignableFrom(List.class);
    }

    /**
     * Guess the property type of field of a class.
     *
     * @param field Field.
     * @return Property type.
     */
    public static PropertyType guessPropertyType(Field field) {
        Class<?> clazz = field.getType();

        if (field.getType().isAssignableFrom(List.class)) {
            clazz = (Class<?>) ((ParameterizedType) field.getGenericType()).getActualTypeArguments()[0];
        }

        if (Number.class.isAssignableFrom(clazz)) {
            return PropertyType.NUMBER;
        } else if (clazz.isAssignableFrom(String.class)) {
            return PropertyType.STRING;
        } else if (clazz.isAssignableFrom(Boolean.class)) {
            return PropertyType.BOOLEAN;
        } else if (clazz.isAssignableFrom(Date.class) || clazz.isAssignableFrom(Instant.class)) {
            return PropertyType.DATE;
        } else if (clazz.isAssignableFrom(KeyValue.class)) {
            return PropertyType.KEY_VALUE;
        }

        // otherwise it must be nested
        return PropertyType.NESTED;
    }

    public static Class<?> getContainedClass(Field field) {
        Class<?> clazz = field.getType();

        if (field.getType().isAssignableFrom(List.class)) {
            clazz = (Class<?>) ((ParameterizedType) field.getGenericType()).getActualTypeArguments()[0];
        }

        return clazz;
    }
}

File: core/src/main/kotlin/cloudspec/annotation/ResourceDefReflectionUtil.kt
/*-
 * #%L
 * CloudSpec Core Library
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.annotation

import arrow.core.Option
import arrow.core.extensions.fx
import arrow.core.none
import arrow.core.toOption
import arrow.syntax.collections.flatten
import cloudspec.model.*
import org.slf4j.LoggerFactory
import java.time.Instant
import java.util.*
import kotlin.reflect.KClass
import kotlin.reflect.KProperty
import kotlin.reflect.KType
import kotlin.reflect.full.*

/**
 * Utility to extract resource definition information out of annotated classes.
 */
object ResourceDefReflectionUtil {
    private val logger = LoggerFactory.getLogger(javaClass)

    /**
     * Extract resource definition out of a class annotated with @ResourceDefinition.
     *
     * @param kclass Class.
     * @return Optional resource definition.
     */
    fun toResourceDef(kclass: KClass<*>): Option<ResourceDef> {
        return if (isValid(kclass)) {
            Option.fx {
                val (ref) = toResourceDefRef(kclass)
                val (description) = toResourceDescription(kclass)

                ResourceDef(ref,
                            description,
                            toPropertyDefs(kclass),
                            toAssociationDefs(kclass))
            }
        } else {
            return none()
        }
    }

    /**
     * Extract resource definition reference out of a class annotated with @ResourceDefinition.
     *
     * @param kclass Class.
     * @return Optional resource definition reference.
     */
    fun toResourceDefRef(kclass: KClass<*>): Option<ResourceDefRef> {
        return kclass.findAnnotation<ResourceDefinition>().toOption().map { annotation ->
            if (logger.isDebugEnabled) {
                logger.debug("Obtained " +
                                     "provider='${annotation.provider}', " +
                                     "group='${annotation.group}', " +
                                     "name='${annotation.name}' " +
                                     "from class ${kclass.qualifiedName}")
            }

            ResourceDefRef(annotation.provider, annotation.group, annotation.name)
        }
    }

    /**
     * Extract resource description out of a class annotated with @ResourceDefinition.
     *
     * @param kclass Class.
     * @return Optional resource description.
     */
    fun toResourceDescription(kclass: KClass<*>): Option<String> {
        return kclass.findAnnotation<ResourceDefinition>().toOption().map { annotation ->
            if (logger.isDebugEnabled) {
                logger.debug("Obtained description='${annotation.description}' " +
                                     "from class ${kclass.qualifiedName}")
            }

            annotation.description
        }
    }

    /**
     * Extract all property definitions out of a class' properties annotated with @PropertyDefinition.
     *
     * @param kclass Class.
     * @return List of property definitions.
     */
    fun toPropertyDefs(kclass: KClass<*>): Set<PropertyDef> {
        return kclass.memberProperties
                .map { toPropertyDef(it) }
                .flatten()
                .toSet()
    }

    /**
     * Extract property definition out of a class' property annotated with @PropertyDefinition.
     *
     * @param kprop Property.
     * @return Optional property definition.
     */
    private fun toPropertyDef(kprop: KProperty<*>): Option<PropertyDef> {
        return Option.fx {
            val (annotation) = kprop.findAnnotation<PropertyDefinition>().toOption()
            val (propertyType) = guessPropertyType(kprop)

            when (propertyType) {
                PropertyType.KEY_VALUE,
                PropertyType.NUMBER,
                PropertyType.STRING,
                PropertyType.BOOLEAN,
                PropertyType.DATE ->
                    PropertyDef(annotation.name,
                                annotation.description,
                                propertyType,
                                isMultiValued(kprop),
                                annotation.exampleValues)
                PropertyType.NESTED -> {
                    val nestedClass = toClass(kprop.returnType)
                    PropertyDef(annotation.name,
                                annotation.description,
                                PropertyType.NESTED,
                                isMultiValued(kprop),
                                annotation.exampleValues,
                                toPropertyDefs(nestedClass),
                                toAssociationDefs(nestedClass))
                }
            }
        }
    }

    /**
     * Extract all association definitions out of properties annotated with @AssociationDefinition.
     *
     * @param kclass Class.
     * @return List of association definitions.
     */
    fun toAssociationDefs(kclass: KClass<*>): Set<AssociationDef> {
        return kclass.memberProperties
                .map { toAssociationDef(kclass, it) }
                .flatten()
                .toSet()
    }

    /**
     * Extract association definition out of a property annotated with @AssociationDefinition.
     *
     * @param kclass Class.
     * @param kprop Property.
     * @return Optional association definition.
     */
    private fun toAssociationDef(kclass: KClass<*>, kprop: KProperty<*>): Option<AssociationDef> {
        val realType = guessContainedType(kprop.returnType)

        // only strings are supported
        if (!isNullableString(realType)) {
            logger.warn("Type $realType of association '${kprop.name}' " +
                                "in class ${kclass.qualifiedName} is not supported")
            return none<AssociationDef>()
        }

        return Option.fx {
            val (annotation) = kprop.findAnnotation<AssociationDefinition>().toOption()
            val (ref) = toResourceDefRef(annotation.targetClass)
            AssociationDef(annotation.name,
                           annotation.description,
                           ref,
                           isMultiValued(kprop))
        }
    }

    /**
     * Guess the property type of a class property.
     *
     * @param kprop Class property.
     * @return Optional property type.
     */
    fun guessPropertyType(kprop: KProperty<*>): Option<PropertyType> {
        return guessPropertyType(kprop.returnType)
    }

    /**
     * Guess the property type of a type.
     *
     * @param ktype Type.
     * @return Optional property type.
     */
    fun guessPropertyType(ktype: KType): Option<PropertyType> {
        return when {
            isMultiValued(ktype) -> guessPropertyType(guessContainedType(ktype))
            isNullableNumber(ktype) -> PropertyType.NUMBER.toOption()
            isNullableString(ktype) -> PropertyType.STRING.toOption()
            isNullableBoolean(ktype) -> PropertyType.BOOLEAN.toOption()
            isNullableDate(ktype) -> PropertyType.DATE.toOption()
            isNullableKeyValue(ktype) -> PropertyType.KEY_VALUE.toOption()
            isNullableNestedProperty(ktype) -> PropertyType.NESTED.toOption()
            else -> null.toOption()
        }
    }

    /**
     * Check whether the class is valid.
     */
    fun isValid(kclass: KClass<*>): Boolean {
        return hasResourceAnnotation(kclass) && hasIdAnnotation(kclass)
    }

    /**
     * Check whether the class has the @ResourceDefinition annotation.
     */
    fun hasResourceAnnotation(kclass: KClass<*>): Boolean {
        return kclass.findAnnotation<ResourceDefinition>() != null
    }

    /**
     * Check whether the class has only one @IdDefinition
     */
    fun hasIdAnnotation(kclass: KClass<*>): Boolean {
        val kprops = kclass.memberProperties
                .filter {
                    it.findAnnotation<IdDefinition>() != null
                }
                .filter {
                    if (it.returnType.isMarkedNullable) {
                        logger.error("Id '${it.name}' property of class '${kclass.qualifiedName}' cannot be nullable.")
                        return@filter false
                    }
                    return@filter true
                }

        return when {
            kprops.isEmpty() ->
                false.also {
                    logger.error("Class '${kclass.qualifiedName}' does not have any ID property")
                }
            kprops.size > 1 ->
                false.also {
                    logger.error("Class '${kclass.qualifiedName}' cannot have more than one ID property")
                }
            else -> true
        }
    }

    /**
     * Check whether a class property is multivalued.
     *
     * @param kprop Class property.
     * @return True if type is multi-valued. False otherwise.
     */
    fun isMultiValued(kprop: KProperty<*>): Boolean {
        return isMultiValued(kprop.returnType)
    }

    /**
     * Check whether a type is multivalued.
     *
     * @param ktype Type.
     * @return True if type is multi-valued. False otherwise.
     */
    fun isMultiValued(ktype: KType): Boolean {
        return isNullableList(ktype) || isNullableSet(ktype)
    }

    /**
     * Check whether a type is a list.
     *
     * @param ktype Type.
     * @return True if type is a list. False otherwise.
     */
    fun isNullableList(ktype: KType): Boolean {
        return ktype.isSubtypeOf(List::class.starProjectedType.withNullability(true))
    }

    /**
     * Check whether a type is a set.
     *
     * @param ktype Type.
     * @return True if type is a set. False otherwise.
     */
    fun isNullableSet(ktype: KType): Boolean {
        return ktype.isSubtypeOf(Set::class.starProjectedType.withNullability(true))
    }

    /**
     * Check whether a type is a number.
     *
     * @param ktype Type.
     * @return True if type is a number. False otherwise.
     */
    fun isNullableNumber(ktype: KType): Boolean {
        return ktype.isSubtypeOf(Number::class.starProjectedType.withNullability(true))
    }

    /**
     * Check whether a type is a string.
     *
     * @param ktype Type.
     * @return True if type is a string. False otherwise.
     */
    fun isString(ktype: KType): Boolean {
        return ktype.isSubtypeOf(String::class.starProjectedType)
    }

    /**
     * Check whether a type is a string.
     *
     * @param ktype Type.
     * @return True if type is a string. False otherwise.
     */
    fun isNullableString(ktype: KType): Boolean {
        return ktype.isSubtypeOf(String::class.starProjectedType.withNullability(true))
    }

    /**
     * Check whether a type is a boolean.
     *
     * @param ktype Type.
     * @return True if type is a boolean. False otherwise.
     */
    fun isNullableBoolean(ktype: KType): Boolean {
        return ktype.isSubtypeOf(Boolean::class.starProjectedType.withNullability(true))
    }

    /**
     * Check whether a type is a date.
     *
     * @param ktype Type.
     * @return True if type is a date. False otherwise.
     */
    fun isNullableDate(ktype: KType): Boolean {
        return ktype.isSubtypeOf(Date::class.starProjectedType.withNullability(true)) ||
                ktype.isSubtypeOf(Instant::class.starProjectedType.withNullability(true))
    }

    /**
     * Check whether a property of a class is a key-value.
     *
     * @param ktype Type.
     * @return True if type is a key-value. False otherwise.
     */
    fun isNullableKeyValue(ktype: KType): Boolean {
        return ktype.isSubtypeOf(KeyValue::class.starProjectedType.withNullability(true))
    }

    /**
     * Check whether a property of a class is a nested property.
     *
     * @param ktype Type.
     * @return True if type is a nested property. False otherwise.
     */
    fun isNullableNestedProperty(ktype: KType): Boolean {
        return (ktype.classifier as KClass<*>).memberProperties.any { it.findAnnotation<PropertyDefinition>() != null }
    }

    /**
     * If a type has generics, return the first type parameter. Return the same type otherwise.
     */
    fun guessContainedType(ktype: KType): KType {
        if (ktype.arguments.isNotEmpty()) {
            return ktype.arguments[0].type ?: ktype
        }
        return ktype
    }

    /**
     * Convert a type to a class.
     *
     * @param ktype Type.
     * @return Class.
     */
    fun toClass(ktype: KType): KClass<*> {
        return guessContainedType(ktype).classifier as KClass<*>
    }
}
